src/transformers/models/bamba/modular_bamba.py

```diff
@@ -949,8 +993,32 @@ def forward(
         output_hidden_states: Optional[bool] = None,
         return_dict: Optional[bool] = None,
         cache_position: Optional[torch.LongTensor] = None,
-        **kwargs,  # NOOP kwargs, for now
+        cu_seq_lens_q: Optional[torch.LongTensor] = None,
+        cu_seq_lens_k: Optional[torch.LongTensor] = None,
+        max_length_q: Optional[int] = None,
+        max_length_k: Optional[int] = None,
+        seq_idx: Optional[torch.IntTensor] = None,
+        **kwargs,
     ) -> Union[Tuple, BaseModelOutputWithPast]:
+        # Padding-free args correctness checks: seq_idx and position_ids must always be provided
+        # for padding-free. FlashAttentionKwargs can optionally be provided, along with position_ids
+        # and seq_idx.
+        flash_attn_kwargs = (cu_seq_lens_q, cu_seq_lens_k, max_length_q, max_length_k)
+        flash_attn_kwargs_all_provided = all(x is not None for x in flash_attn_kwargs)
+        flash_attn_kwargs_none_provided = all(x is None for x in flash_attn_kwargs)
+        if not (flash_attn_kwargs_all_provided or flash_attn_kwargs_none_provided):
+            raise ValueError(
+                "Either all of (cu_seq_lens_q, cu_seq_lens_k, max_length_q, max_length_k)"
+                " must be None, or they must all be provided."
+            )
+        if flash_attn_kwargs_all_provided and (position_ids is None or seq_idx is None):
+            raise ValueError(
+                "If (cu_seq_lens_q, cu_seq_lens_k, max_length_q, max_length_k) are provided,"
+                " then position_ids and seq_idx must also be provided."
+            )
```

Comment:
```suggestion
        if flash_attn_kwargs_all_provided and position_ids is None:
            raise ValueError(
                "If (cu_seq_lens_q, cu_seq_lens_k, max_length_q, max_length_k) are provided,"
                " then position_ids must also be provided."
            )
```
Could we not have fa kwarg and position ids? This would enable fa path but not the mamba2 path - which shouldn't be an issue imo.

Might be nice to add for which path here.

Reply:
Ah, I guess that's possible if the `cu_seq_lens` are trivial like `(0, seq_len)`, yeah.

The concern I had here is silent incorrectness if non-trival `cu_seq_lens = (0, seq_len0, seq_len0 + seq_len1, ...)` are passed in without a corresponding non-trivial `seq_idx`, in which case the outputs would be silently incorrect.

I'll sharpen up the checks here, thanks.

Reply:
Argh yea that's painful :/

Reply:
No worries take your time :)